### PR TITLE
client: improve error from client.do() on json decode failures

### PR DESCRIPTION
**Pull request by mvo5**

### CODE REVIEW

**Comment by fgimenez** [comment]
LVGTM

**Comment by stevenwilkin** [comment]
👍

**Comment by mvo5** [comment]
Hm, actually - rsp.Body is a io.Reader so this change won't help to add clarity :( Closing for now.

File: client/client.go [hunk 1]
```diff
@@ -157,7 +158,12 @@ func (client *Client) do(method, path string, query url.Values, headers map[stri
 	if v != nil {
 		dec := json.NewDecoder(rsp.Body)
 		if err := dec.Decode(v); err != nil {
-			return err
+			r := dec.Buffered()
+			buf, err1 := ioutil.ReadAll(r)
+			if err1 != nil {
+				buf = []byte(fmt.Sprintf("error reading buffered: %s", err1))
+			}
+			return fmt.Errorf("cannot decode %q: %s", string(buf), err)
```
Comment (pedronis) [inline client/client.go]: string() is not needed, "%q* works fine with []byte

File: client/client.go [hunk 2]
```diff
@@ -157,7 +158,12 @@ func (client *Client) do(method, path string, query url.Values, headers map[stri
 	if v != nil {
 		dec := json.NewDecoder(rsp.Body)
 		if err := dec.Decode(v); err != nil {
-			return err
+			r := dec.Buffered()
+			buf, err1 := ioutil.ReadAll(r)
+			if err1 != nil {
+				buf = []byte(fmt.Sprintf("error reading buffered: %s", err1))
```
Comment (pedronis) [inline client/client.go]: buffered response body?

**Pull request merged**